### PR TITLE
Intersection of two

### PR DESCRIPTION
**Pull request by linahKhayri**
---
name: solution review of intersection_of_two function
about: A template PR for code review with a checklist
---

 [here's the issue](https://github.com/MIT-Emerging-Talent/ET6-foundations-group-24/issues/53)

This function takes two lists and returns a list containing the intersection of them.
    The intersection includes only the items that are present in both
    list1 and list2. The items in the intersection are returned only once,
    even if they appear multiple times in the input lists. The input lists
    can contain any type of items that are allowed in a list (e.g., integers,
    strings, floats, special characters, etc.)

    Note:
        The function could be simplified using built-in functions like set(),
        which automatically handle duplicates and set intersections.
        However, I wanted to write the logic manually to demonstrate how it works step-by-step

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Review by robiel0143 — APPROVED** [review]
Well prepared.

File: solutions/tests/test_intersection_of_two.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        second_list2 = []
+        with self.assertRaises(AssertionError):
+            intersection_of_two(first_list, second_list2)
+
+    def test_any_non_list_input(self):
+        """Test when any of the inputs is not a list."""
+        first_list = "cat"
+        second_list2 = []
+        with self.assertRaises(AssertionError):
+            intersection_of_two(first_list, second_list2)
```
Comment (ViktoriyaHaiduk) [inline solutions/tests/test_intersection_of_two.py:147]: Great work! Everything is ok!

**Review by ViktoriyaHaiduk — DISMISSED** [review]
Great work! Everything is ok!

**Comment by linahKhayri** [comment]
> Great work! Everything is ok!

Thank you so much @ViktoriyaHaiduk for the feedback 💖
Can you please do the checks in the pr checklist?

File: solutions/intersection_of_two.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    """
+    assert isinstance(first_list, list), "first list must be a list"
+    assert isinstance(second_list, list), "second list must be a list"
+    assert len(first_list) > 0, "list1 can't be empty"
+    assert len(second_list) > 0, "list2 can't be empty"
+
+    intersection_list = []
+    # Loop through each item in the first list
+    for item in first_list:
```
Comment (Mohamed-Elnageeb) [inline solutions/intersection_of_two.py:61]: I like the strategy behind your solution. However, the implementation could potentially be optimized for better performance by using sets instead of lists. That said, this change is not strictly necessary

File: solutions/intersection_of_two.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    [3, 4]
+    >>> intersection_of_two(['a', 'cat', 'ABC', 'dog'],['c', 'abc', 'cat'])
+    ['cat']
+    >>> intersection_of_two(['$', '&', '@'],['$', '&', '@'])
+    ['$', '&', '@']
+    >>> intersection_of_two(['1', 3, 'girl'],['a', 4, 3, '1', 'boy'])
+    ['1', 3]
+
+    """
+    assert isinstance(first_list, list), "first list must be a list"
```
Comment (Mohamed-Elnageeb) [inline solutions/intersection_of_two.py:54]: here in the comments i think you should write the name of the variable to make the meaning more clear : 
i.e first_list must be a list

File: solutions/tests/test_intersection_of_two.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    - Standard cases: Lists with common items, lists with special characters.,
+        lists with mixed data types,
+
+
+    - Edge cases: Lists with no intersection, single-element lists, duplicated items,
+        different types but equal values, lists with spaces items,
+        sublist intersection, different case items
+
+    - Defensive tests: Non-list inputs, one empty list, both empty lists
```
Comment (Mohamed-Elnageeb) [inline solutions/tests/test_intersection_of_two.py:20]: good well-rounded tests

**Review by Mohamed-Elnageeb — CHANGES_REQUESTED** [review]
Good clean code linah! 
only small changes needed.

File: solutions/intersection_of_two.py (hunk 2 above)
Comment (linahKhayri) [reply solutions/intersection_of_two.py:61]: Thank you Mohamed,🙏
Indeed it may give faster results but I included this in the function docstring as a note: `(This function could be simplified using built-in functions like set(), 
        which automatically handle duplicates and set intersections.
        However, I wanted to write the logic manually to demonstrate how it works step-by-step) `
So I used a loop instead of the built in set() function just to show how the process works step-by-step. But if you're talking about using sets as inputs instead of lists, that's actually a great idea and might be the better choice but I just wanted to manually handle the duplicate elements and show how they get executed once.

File: solutions/intersection_of_two.py (hunk 3 above)
Comment (linahKhayri) [reply solutions/intersection_of_two.py]: Done, thankss 💖

File: solutions/intersection_of_two.py (hunk 2 above)
Comment (Mohamed-Elnageeb) [reply solutions/intersection_of_two.py:61]: oh ok i missed that one! good 👍

**Review by Mohamed-Elnageeb — APPROVED** [review]

**Comment by linahKhayri** [comment]
> Good clean code linah! only small changes needed.

Thank you so much, I really appreciate the feedback 💖
